pkg/router/template/router.go

```diff
@@ -276,6 +277,79 @@ func secretToPem(secPath, outName string) error {
 	return ioutil.WriteFile(outName, pemBlock, 0444)
 }
 
+// watchSecretDir adds a watcher on the input directory that updates the output
+// PEM file when a change is detected.
+func (r *templateRouter) watchSecretDir(secPath, outName string) error {
```

Comment:
Throughout: instead of logs in the main loop path, how about capturing metrics?

Reply:
Do you want a new metric in addition to the `metricReload` and `metricWriteConfig` metrics?

Reply:
Seems like having more metrics is generally useful. Is there a disadvantage to adding more metric types here?

Reply:
It seems strange to have a metric for this specific case.  Do we want a metric for each reload trigger?

Reply:
For a basis of comparison, what questions do the existing logging here answer, and who's asking them?

A generic benefit of metrics is they doesn't presuppose the questions — metrics provide the ability to formulate ad-hoc lines of questioning.

Here are some questions:

1. How often are we reloading the router over <timespan>
2. How often are we refreshing the certificate secret over <timespan>
3. Are we refreshing the certificate abnormally (imagine an fsnotify issue — notice the lack of rate limiting in this code)

More metrics points also enable new alerts to be imagined. For example, we could even automate question 3 by defining an alert to trigger when the secret refreshes more than twice a minute.

Events would also have advantages over logs here, but I'm not sure if they're worth it?

Reply:
> For a basis of comparison, what questions do the existing logging here answer, and who's asking them?

"Did the router notice that I updated the default certificate secret?", or, "Why isn't the router using the default certificate that I just configured?".

> A generic benefit of metrics is they doesn't presuppose the questions — metrics provide the ability to formulate ad-hoc lines of questioning.
> 
> Here are some questions:
> 
> 1. How often are we reloading the router over _timespan_
> 2. How often are we refreshing the certificate secret over _timespan_
> 3. Are we refreshing the certificate abnormally (imagine an fsnotify issue — notice the lack of rate limiting in this code)
> 
> More metrics points also enable new alerts to be imagined. For example, we could even automate question 3 by defining an alert to trigger when the secret refreshes more than twice a minute.

I think the two interesting case are (1) the secret was changed or (2) the watch broke, for the following scenarios:

* The administrator changed the default certificate and expects feedback that the change has been recognized—case (1).
* The administrator did not change the default certificate and should be alerted if a change has been recognized—absence of case (1).
* The watch broke—case (2)—which would could indicate that the platform is unstable.

> Events would also have advantages over logs here, but I'm not sure if they're worth it?

An event seems appropriate for (1) above.  Can one configure an alert for an event?

For case (2), it almost seems reasonable to crashloop... But maybe we could emit an event and recreate the watch.